server/src/main/java/org/candlepin/hostedtest/HostedTestSubscriptionServiceAdapter.java

```diff
@@ -289,6 +289,8 @@ public ProductInfo createProduct(ProductInfo pinfo) {
         pdata.setCreated(new Date());
         pdata.setUpdated(new Date());
         pdata.setBranding(this.resolveBranding(pinfo.getBranding()));
+        pdata.setProvidedProducts(this.resolveProvidedProducts(pinfo.getProvidedProducts()));
```

Comment:
I don't think the new method `resolveProvidedProducts` is needed. We could simply use the existing `resolveProduct/resolveProducts` methods to lookup any existing provided products we have, assuming they were already created.

spec tests would have to make sure that the provided products are created first, and then create a marketing product that references them.

The only reason the `resolveBranding` method was created this way was because you can't really create a branding standalone. It is an entity fully owned by a product.

Reply:
This was done to resolve `'verify branding info is correct in json blob'` spec test. Provided product is created in non-hosted mode. When the marketing products get created in hosted mode, it doesn't recognize provided product in `productMap` of HostedTestSubscriptionServiceAdapter.java. Hence for the solution, I added `resolveProvidedProducts` method. 

Possible approaches :
1. As suggested, modify the spec test to create provided product before marketing product in hosted.
2. Keeping `resolveProvidedProducts` without identifying either the provided product is created in hosted or not.

Reply:
We should use approach 1 because it is consistent with how the HostedTest adapter works so far. That test already has an `if is_hosted?` condition; you can use that to create the engineering product either in hosted or in normal mode.

Reply:
Did changes as per approach 1.